Canvas should be able to draw a WebCodecsVideoFrame

youennf opened the pull request:
#### ed683c9b20ac22a039a09d22efd38cafb057c4bb
<pre>
Canvas should be able to draw a WebCodecsVideoFrame
<a href="https://bugs.webkit.org/show_bug.cgi?id=246392">https://bugs.webkit.org/show_bug.cgi?id=246392</a>
rdar://problem/101066016

Reviewed by Eric Carlson.

Make sure to create a correct WebCodecsVideoFrame from canvas.
Update IDL and binding code to allow taking WebCodecsVideoFrame as a CanvasImageSource.
Add prelimianry support to paint a WebCodecsVideoFrame in a canvas.
Add specific expectations since display-p3 is not defined in iOS simulator and BigSur.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource.html:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/webcodecs/videoFrame-drawImage.any-expected.txt: Added.
* LayoutTests/platform/mac-bigsur-wk2/imported/w3c/web-platform-tests/webcodecs/videoFrame-drawImage.any-expected.txt: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::checkImageUsability):
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameFromOtherFrame):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
(WebCore::WebCodecsVideoFrame::shoudlDiscardAlpha const):
* Source/WebCore/html/canvas/CanvasDrawImage.idl:
* Source/WebCore/html/canvas/CanvasFillStrokeStyles.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::size):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::createPattern):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::paintInContext):
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::paintVideoFrame):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::paintInContext):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::paintVideoFrame):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:

Canonical link: <a href="https://commits.webkit.org/255720@main">https://commits.webkit.org/255720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/542d9aa3345ce852ccb2675ae1e9c167a7a748b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/93429 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/2625 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/24073 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103101 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/163431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/97427 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/2635 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/30927 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/85797 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99202 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99092 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/1853 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/79880 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/28901 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/83766 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/83515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/71857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37309 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/17382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35139 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/18632 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3956 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39012 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/41159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/40948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/37857 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
```diff
@@ -1514,6 +1528,23 @@ static std::pair<FloatRect, FloatRect> normalizeSourceAndDestination(const Float
+#if ENABLE(WEB_CODECS)
+ExceptionOr<void> CanvasRenderingContext2DBase::drawImage(WebCodecsVideoFrame& frame, const FloatRect&, const FloatRect& dstRect)
+{
+    auto* c = drawingContext();
```
`c` could be more descriptive.

youennf replied:
Will change

eric-carlson commented on Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
```diff
@@ -344,6 +346,17 @@
+    // FIXME: Handle alpha discarding.
+    UNUSED_PARAM(shouldDiscardAlpha);
+
+    auto conformer = makeUnique<PixelBufferConformerCV>((__bridge CFDictionaryRef)@{ (__bridge NSString *)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32BGRA) });
```
It is really unfortunate you have to create a conformer every time a frame is drawn to a context.

youennf replied:
Right, in practice, I believe the code path will be different since painting should happen in GPUProcess.
I'll add a FIXME, maybe the context should own its conformer.

eric-carlson reviewed (approved):
r=me

youennf committed:
Canvas should be able to draw a WebCodecsVideoFrame
https://bugs.webkit.org/show_bug.cgi?id=246392
rdar://problem/101066016

Reviewed by Eric Carlson.

Make sure to create a correct WebCodecsVideoFrame from canvas.
Update IDL and binding code to allow taking WebCodecsVideoFrame as a CanvasImageSource.
Add prelimianry support to paint a WebCodecsVideoFrame in a canvas.
Add specific expectations since display-p3 is not defined in iOS simulator and BigSur.

* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource.html:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/webcodecs/videoFrame-drawImage.any-expected.txt: Added.
* LayoutTests/platform/mac-bigsur-wk2/imported/w3c/web-platform-tests/webcodecs/videoFrame-drawImage.any-expected.txt: Added.
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.cpp:
(WebCore::checkImageUsability):
(WebCore::WebCodecsVideoFrame::create):
(WebCore::WebCodecsVideoFrame::initializeFrameFromOtherFrame):
* Source/WebCore/Modules/webcodecs/WebCodecsVideoFrame.h:
(WebCore::WebCodecsVideoFrame::shoudlDiscardAlpha const):
* Source/WebCore/html/canvas/CanvasDrawImage.idl:
* Source/WebCore/html/canvas/CanvasFillStrokeStyles.idl:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::size):
(WebCore::CanvasRenderingContext2DBase::drawImage):
(WebCore::CanvasRenderingContext2DBase::createPattern):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.h:
* Source/WebCore/inspector/InspectorCanvas.cpp:
(WebCore::InspectorCanvas::processArgument):
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
* Source/WebCore/platform/VideoFrame.cpp:
(WebCore::VideoFrame::paintInContext):
* Source/WebCore/platform/VideoFrame.h:
* Source/WebCore/platform/graphics/GraphicsContext.cpp:
(WebCore::GraphicsContext::paintVideoFrame):
* Source/WebCore/platform/graphics/GraphicsContext.h:
* Source/WebCore/platform/graphics/cv/VideoFrameCV.mm:
(WebCore::VideoFrame::paintInContext):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.cpp:
(WebCore::DisplayList::Recorder::paintVideoFrame):
* Source/WebCore/platform/graphics/displaylists/DisplayListRecorder.h:

Canonical link: https://commits.webkit.org/255720@main
